ctapipe/image/extractor.py

```diff
@@ -1287,18 +1289,127 @@ def __call__(
+        integral = ref_pulse_shape.sum() * ref_sample_width / camera_sample_width
+        phase_gains, phase_shifts = [], []
+        for phase in range(avg_step):
+            x = ref_pulse_shape[phase::avg_step]
+            y = deconvolve(np.atleast_2d(x), 0.0, upsampling, pz)[0]
+            i_max = y.argmax()
+            start = i_max - window_shift  # TODO should use extract_around_peak here
+            stop = start + window_width
+            if start >= 0 and stop <= y.size:
+                phase_shifts.append((i_max / upsampling * avg_step - ref_pulse_shape.argmax()) * ref_sample_width)
+                phase_gains.append(y[start:stop].sum() / integral)
+
+        assert(len(phase_gains))
+        gains.append(np.mean(phase_gains))
+        shifts.append(np.mean(phase_shifts))
+
+    return pzs, gains, shifts
+
+
+def deconvolve(waveforms, bls, up : int, pz : float):
+    assert(len(waveforms.shape) == 2)
+    bls = np.atleast_2d(bls).T
+    y = waveforms - bls
+    y[:,1:] -= pz * y[:,:-1]
+    y[:,0] = 0
+    if up > 1:
+        return filtfilt(np.ones(up), up, np.repeat(y, up, axis=-1))
+    return y
 
 
 class FlashCamExtractor(ImageExtractor):
-    def __call__(self, waveforms, telid, selected_gain_channel) -> DL1CameraContainer:
+    upsampling = IntTelescopeParameter(
+        default_value=4, help="Define the upsampling factor for waveforms"
+    ).tag(config=True)
 
-        return DL1CameraContainer(
-            image=np.zeros(...),
-            peak_time=...,
-            is_valid=True,
+    window_width = IntTelescopeParameter(
+        default_value=4, help="Define the width of the integration window"
+    ).tag(config=True)
+
+    window_shift = IntTelescopeParameter(
+        default_value=2,
+        help="Define the shift of the integration window from the peak_index "
+        "(peak_index - shift)",
+    ).tag(config=True)
+
+    apply_integration_correction = BoolTelescopeParameter(
+        default_value=True, help="Apply the integration window correction"
+    ).tag(config=True)
+
+    local_weight = IntTelescopeParameter(
+        default_value=0,
+        help="Weight of the local pixel (0: peak from neighbors only, "
+        "1: local pixel counts as much as any neighbor)",
+    ).tag(config=True)
+
+    def __init__(self, subarray, **kwargs):
+        super().__init__(subarray=subarray, **kwargs)
+
+        self.sampling_rate_ghz = {
+            tel_id: telescope.camera.readout.sampling_rate.to_value("GHz")
+            for tel_id, telescope in subarray.tel.items()
+        }
+
+    def __call__(self, waveforms, tel_id, _, broken_pixels) -> DL1CameraContainer:
+        upsampling = max(1, self.upsampling.tel[tel_id])
+        window_width = max(1, self.window_width.tel[tel_id])
+        window_shift = self.window_shift.tel[tel_id]
+        pzs, gains, shifts = deconvolution_parameters(self.subarray.tel[tel_id].camera, upsampling, window_width, window_shift)
```

Comment:
The call to `deconvolution_parameters` looks expensive (didn't test) but does not change since it only depends on options of the Extractor and on the subarray, so this could also be done only once in `__init__`

Reply:
I was thinking about that - but wasn't sure if the parameters (upsampling, window_width, window_size) are *fixed* after init. Is that the case?

Reply:
Yes, in most cases we assume that configuration cannot change after `__init__`. 

The only way that can happen is in interactive settings like jupyter notebooks, but then you should just run `__init__` again.

Reply:
OK if such behaviour is acceptable - done. 👍